src/BlockTypes/ProductCollection.php

```diff
@@ -326,4 +351,94 @@ private function array_merge_recursive_replace_non_array_properties( $base, $new
 
 		return $base;
 	}
+
+	/**
+	 * Return a query for products depending on their stock status.
+	 *
+	 * @param array $stock_statii An array of acceptable stock statii.
+	 * @return array
+	 */
+	private function get_stock_status_query( $stock_statii ) {
+		if ( ! is_array( $stock_statii ) ) {
+			return array();
+		}
+
+		$stock_status_options = array_keys( wc_get_product_stock_status_options() );
+
+		/**
+		 * If all available stock status are selected, we don't need to add the
+		 * meta query for stock status.
+		 */
+		if (
+			count( $stock_statii ) === count( $stock_status_options ) &&
+			array_diff( $stock_statii, $stock_status_options ) === array_diff( $stock_status_options, $stock_statii )
+		) {
+			return array();
+		}
+
+		/**
+		 * If all stock statuses are selected except 'outofstock', we use the
+		 * product visibility query to filter out out of stock products.
```

Comment:
I have to say that, as a user, I would find this a bit confusing because the block settings and WC settings can contradict each other. IMO block settings should take priority over WooCommerce settings. So if the _Hide out of stock items from the catalog_ option is checked but I add a Product Collection block selecting to display products with all stock statuses, I would expect all products to appear, ignoring the _Hide out of stock items from the catalog_ option.

In other words, I think the _Hide out of stock items from the catalog_ option should affect which stock status options are selected by default when inserting the block, but they shouldn't affect how the block works once the block attributes have been set.

But... I don't have a strong opinion so I'm happy to change my mind. Also, I don't think it's a blocker.

Reply:
Hey @Aljullu,

> I think the Hide out of stock items from the catalog option should affect which stock status options are selected by default when inserting the block, but they shouldn't affect how the block works once the block attributes have been set.

I just want to confirm that currently, when we insert the block, we do take the "Hide out of stock items from the catalog" setting into consideration. This is implemented [here](https://github.com/woocommerce/woocommerce-blocks/blob/0c30f5b51c4594654ffe5dbb0e937f83f623bde8/assets/js/blocks/product-collection/constants.ts#L18-L32).

Also, I agree that it makes sense to ignore the "Hide out of stock items from the catalog" setting if we allow merchants to explicitly set stock statuses in the Product Collection block.
![image](https://github.com/woocommerce/woocommerce-blocks/assets/16707866/166961ab-ffd8-431c-aed9-92b90fbf3f5b)


PS: I will go ahead and merge this PR for now. We can continue our discussion here. If we decide to ignore the global "Hide out of stock items from the catalog" setting, I will create a new PR for that.

CC: @nerrad @tjcafferkey

Reply:
> I just want to confirm that currently, when we insert the block, we do take the "Hide out of stock items from the catalog" setting into consideration. This is implemented [here](https://github.com/woocommerce/woocommerce-blocks/blob/0c30f5b51c4594654ffe5dbb0e937f83f623bde8/assets/js/blocks/product-collection/constants.ts#L18-L32).

Oh right, I missed it! That's nice.